power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -750,6 +847,77 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
         return tap_changed;
     }
 
+    bool adjust_transformer_bs(RegulatedTransformer const& regulator, State const& state,
+                               ResultType const& solver_output, UpdateBuffer& update_data, bool strategy_max,
+                               std::size_t i, std::size_t j) const {
+        bool tap_changed = false;
+        auto& bs_ref = binary_search_[i][j];
+
+        regulator.transformer.apply([&](transformer_c auto const& transformer) {
+            using TransformerType = std::remove_cvref_t<decltype(transformer)>;
+            using sym = typename ResultType::value_type::sym;
+
+            auto const param = regulator.regulator.get().template calc_param<sym>();
+            auto const node_state =
+                NodeState<sym>{.u = u_pu_controlled_node<TransformerType>(regulator, state, solver_output),
+                               .i = i_pu_controlled_node<TransformerType>(regulator, state, solver_output)};
+
+            // Only return false if the binary search condition is met
+            if (bs_ref.get_end_of_bs() || bs_ref.get_bs_inevitable_run()) {
+                tap_changed = false;
+                return;
+            }
+
+            auto const cmp = node_state <=> param;
+            auto new_tap_pos = [&cmp, strategy_max, &bs_ref] {
+                if (cmp != 0) {                       // NOLINT(modernize-use-nullptr)
+                    auto state_above_range = cmp > 0; // NOLINT(modernize-use-nullptr)
+                    auto is_down = state_above_range ? bs_ref.get_bs_tap_reverse() : !bs_ref.get_bs_tap_reverse();
+                    if (bs_ref.get_bs_last_check()) {
+                        bs_ref.set_bs_current_tap(is_down ? bs_ref.get_bs_tap_left() : bs_ref.get_bs_tap_right());
+                        bs_ref.set_bs_inevitable_run(true);
+                        return bs_ref.get_bs_current_tap();
+                    }
+                    bs_ref.set_bs_last_down(is_down);
+                    bs_ref.adjust_bs(strategy_max);
+                }
+                return bs_ref.get_bs_current_tap();
+            }();
+
+            if (new_tap_pos != transformer.tap_pos()) {
+                bs_ref.set_bs_current_tap(new_tap_pos);
+                add_tap_pos_update(new_tap_pos, transformer, update_data);
+                tap_changed = true;
+                return;
+            }
+
+            const bool previous_down = bs_ref.get_bs_last_down();
+            if (bs_ref.get_bs_tap_reverse() ? strategy_max : !strategy_max) {
+                bs_ref.set_bs_tap_left(bs_ref.get_bs_current_tap());
+                bs_ref.set_bs_last_down(false);
+            } else {
+                bs_ref.set_bs_tap_right(bs_ref.get_bs_current_tap());
+                bs_ref.set_bs_last_down(true);
+            }
+
+            const bool prefer_higher = strategy_max ? !bs_ref.get_bs_tap_reverse() : bs_ref.get_bs_tap_reverse();
```

Comment:
This logic can be moved in BinarySearch too to make this function nicer.
It seems to only use the BinarySearch object.
It would ofcourse depend on the generalization for the binary search in plan.

Reply:
Can ofcourse be handled later in the PR since this is a preliminary review.

Reply:
Tentatively addressed in the latest commit.